src/platform/audio/ffmpegaudiofilecache.cpp

```diff
@@ -0,0 +1,449 @@
+        captainslog_error("Audio codec not supported: '%u'", file->fmt_ctx->streams[0]->codecpar->codec_tag);
+        Close_FFmpeg_Contexts(file);
+        return false;
+    }
+
+    ret = avcodec_open2(file->fmt_ctx->streams[0]->codec, input_codec, NULL);
+    if (ret < 0) {
+        captainslog_error("Failed to open input codec");
+        Close_FFmpeg_Contexts(file);
+        return false;
+    }
+
+    file->codec_ctx = file->fmt_ctx->streams[0]->codec;
+    return true;
+}
+
+/**
+ * Decode the input data and append it to our wave data stream
+ */
+bool FFmpegAudioFileCache::Decode_FFmpeg(FFmpegOpenAudioFile *file)
+{
+    AVPacket *packet = av_packet_alloc();
+    AVFrame *frame = av_frame_alloc();
+
+    int result = 0;
+    while (av_read_frame(file->fmt_ctx, packet) >= 0) {
+        result = avcodec_send_packet(file->codec_ctx, packet);
+        if (result < 0) {
+            captainslog_error("Failed to send audio packet to decoder.");
+            return false;
+        }
+        result = avcodec_receive_frame(file->codec_ctx, frame);
+        // Check if this was a real error or we just need more data
+        if (result < 0 && result != AVERROR(EAGAIN) && result != AVERROR_EOF) {
+            captainslog_error("Failed to receive audio frame from decoder.");
+            return false;
+        } else if (result >= 0) {
+            int frame_data_size = av_samples_get_buffer_size(
+                NULL, file->codec_ctx->channels, frame->nb_samples, file->codec_ctx->sample_fmt, 1);
+            av_realloc(file->wave_data, file->data_size + frame_data_size);
+            memcpy(file->wave_data + file->data_size, frame->data[0], frame_data_size);
+            file->data_size += frame_data_size;
+        }
+        av_packet_unref(packet);
+    }
+
+    av_packet_free(&packet);
+    av_frame_free(&frame);
+
+    return true;
+}
+
+/**
+ * Close all the open FFmpeg handles for an open file.
+ */
+void FFmpegAudioFileCache::Close_FFmpeg_Contexts(FFmpegOpenAudioFile *file)
+{
+    if (file->fmt_ctx) {
+        avformat_close_input(&file->fmt_ctx);
+    }
+
+    if (file->avio_ctx) {
+        av_freep(&file->avio_ctx);
+    }
+}
+
+void FFmpegAudioFileCache::Fill_Wave_Data(FFmpegOpenAudioFile *file)
+{
+    WavHeader wav;
+    wav.chunk_size = file->data_size - 8;
+    wav.subchunk2_size = file->data_size - 44;
+    wav.channels = file->codec_ctx->channels;
+    wav.bits_per_sample = av_get_bits_per_sample(file->codec_ctx->codec_id);
+    wav.samples_per_sec = file->codec_ctx->sample_rate;
+    wav.bytes_per_sec = file->codec_ctx->sample_rate * file->codec_ctx->channels * (wav.bits_per_sample / 8);
+    wav.block_align = file->codec_ctx->channels * (wav.bits_per_sample / 8);
+    memcpy(file->wave_data, &wav, sizeof(WavHeader));
+}
+
+/**
+ * Opens an audio file. Reads from the cache if available or loads from file if not.
+ */
+uint8_t *FFmpegAudioFileCache::Open_File(const Utf8String &filename)
+{
+    ScopedMutexClass lock(&m_mutex);
```

Comment:
This looks scary, because the entire file read is locked. This means huge performance penalty in threaded context. What exactly needs locking here?

Reply:
I copied that from MilesAudioFileCache